### PR TITLE
Introduce ANTLR parsing for ProgramIndicators

### PR DESCRIPTION
**Pull request by jimgrace**

### CODE REVIEW

File: dhis-2/.gitignore [hunk 1]
```diff
@@ -8,7 +8,7 @@
 /dhis-services/dhis-service-administration/target/
 /dhis-services/dhis-service-analytics/target/
 /dhis-services/dhis-service-core/target/
-/dhis-services/dhis-service-expression-parser/target/
+/dhis-services/dhis-service-parser/target/
```
Comment (larshelge) [inline dhis-2/.gitignore]: Remove.

File: dhis-2/dhis-services/dhis-service-analytics/src/test/java/org/hisp/dhis/analytics/data/AnalyticsServiceMetadataTest.java [hunk 2]
```diff
@@ -129,61 +129,61 @@ public void metadataContainsIndicatorGroupMetadata()
...
-                    ) ) ) )
-            .withIgnoreLimit( true )
-            .withSkipData( true )
-            .build();
+                // DATA ELEMENTS
+                .withDimensions( Lists.newArrayList(
+                        new BaseDimensionalObject( "pe", DimensionType.PERIOD, periods ),
+                        new BaseDimensionalObject( "dx", DimensionType.DATA_X, DISPLAY_NAME_DATA_X,
+                                "display name",
+                                new DimensionalKeywords( Collections.singletonList( indicatorGroup ) ),
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-analytics/src/test/java/org/hisp/dhis/analytics/data/AnalyticsServiceMetadataTest.java]: Ensure 4 spaces for indentation, not 8.

File: dhis-2/dhis-services/dhis-service-analytics/src/test/java/org/hisp/dhis/analytics/data/AnalyticsServiceMetadataTest.java [hunk 3]
```diff
@@ -84,38 +84,38 @@ public void setUp()
...
-                        new OrganisationUnit( "bbb", "bbb", "OU_2", null, null, "c2" )
-                ) ) )
-            )
-            .build();
+                // PERIOD
+                .withPeriod( new Period( YearlyPeriodType.getPeriodFromIsoString( "2017W10" ) ) )
+                // DATA ELEMENTS
+                .withDataElements( newArrayList( createDataElement( 'A', new CategoryCombo() ) ) ).withIgnoreLimit( true )
+                // FILTERS (OU)
+                .withFilters( Collections.singletonList(
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-analytics/src/test/java/org/hisp/dhis/analytics/data/AnalyticsServiceMetadataTest.java]: 4 space indentation.

File: dhis-2/dhis-support/pom.xml [hunk 4]
```diff
@@ -22,6 +22,7 @@
     <module>dhis-support-system</module>
     <module>dhis-support-jdbc</module>
     <module>dhis-support-amqp</module>
+    <module>dhis-support-parser</module>
```
Comment (larshelge) [inline dhis-2/dhis-support/pom.xml]: Should we perhaps call the project `dhis-support-expression-parser` as there might be other "parsers" in the future?

Comment (jimgrace) [reply dhis-2/dhis-support/pom.xml]: Done.

File: dhis-2/.gitignore (hunk 1 above)
Comment (jimgrace) [reply dhis-2/.gitignore]: Fixed (no more service...parser). Also added /dhis-support/dhis-support-expression-parser/target/.

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/ProgramValidator.java [hunk 5]
```diff
@@ -0,0 +1,428 @@
...
+import static org.hisp.dhis.parser.expression.antlr.ExpressionParser.*;
+
+/**
+ * Visit the ANTLR parse tree for a program expression, using the ANTLR visitor
+ * pattern, for syntax checking and also generating an expression description.
+ *
+ * @author Markus Bekken
+ * @author Jim Grace
+ */
+public class ProgramValidator
```
Comment (luciano-fiandesio) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/program/ProgramValidator.java]: Code coverage for this class is around 60%, with lots of private methods with 0% coverage.

File: dhis-2/dhis-services/dhis-service-reporting/src/main/java/org/hisp/dhis/validation/DataValidationTask.java [hunk 6]
```diff
@@ -463,7 +463,7 @@ private void addToAocCache( CategoryOptionCombo aoc )
                 values.putAll( nonAocValues );
             }
 
-            Double value = expressionService.getExpressionValue( expression, values,
```
Comment (luciano-fiandesio) [inline dhis-2/dhis-services/dhis-service-reporting/src/main/java/org/hisp/dhis/validation/DataValidationTask.java]: This class has 0% test coverage

File: dhis-2/dhis-support/dhis-support-expression-parser/src/main/java/org/hisp/dhis/parser/expression/CommonSqlGenerator.java [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+                return args.get( 0 ) +
+                    " <= " + args.get( 1 );
+
+            default:
+                throw new InternalParserException( "Expecting operator, found " + ctx.op.getText() );
+        }
+    }
+
+    @Override
+    public String visitFunction( FunctionContext ctx )
```
Comment (luciano-fiandesio) [inline dhis-2/dhis-support/dhis-support-expression-parser/src/main/java/org/hisp/dhis/parser/expression/CommonSqlGenerator.java]: This function has 0% test coverage

File: dhis-2/dhis-support/dhis-support-expression-parser/src/main/java/org/hisp/dhis/parser/expression/CommonSqlGenerator.java [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+ */
+public abstract class CommonSqlGenerator
+    extends AbstractVisitor
+{
+    // -------------------------------------------------------------------------
+    // Visitor methods
+    // -------------------------------------------------------------------------
+
+    @Override
+    public String visitOperator( ExprContext ctx )
```
Comment (luciano-fiandesio) [inline dhis-2/dhis-support/dhis-support-expression-parser/src/main/java/org/hisp/dhis/parser/expression/CommonSqlGenerator.java]: This function has very low test coverage

File: dhis-2/dhis-support/dhis-support-expression-parser/src/main/java/org/hisp/dhis/parser/expression/ParserUtils.java [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+
+    /**
+     * Casts object as Boolean, or throws exception.
+     * <p/>
+     * If the object is null, returns a default.
+     *
+     * @param object the value to cast as a Boolean.
+     * @return Boolean value.
+     */
+    public static Boolean castBooleanDefault( Object object )
```
Comment (luciano-fiandesio) [inline dhis-2/dhis-support/dhis-support-expression-parser/src/main/java/org/hisp/dhis/parser/expression/ParserUtils.java]: This method seems to be unused

**Pull request merged**